### ISSUE TITLE
Convert "Recent comments" block to a view

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#330]
Sub-issue of https://github.com/backdrop/backdrop-issues/issues/151.

The "Recent comments" block should be converted to a view. In the upgrade path, the "comment_block_count" variable should be used as the default limit on the view, then the variable should be deleted.

---
PR by @klonos: https://github.com/backdrop/backdrop/pull/1693

**Comment by quicksketch** [comment]
Let's move this out of the novice queue. Providing the upgrade path can be quite challenging. See node_update_1008() for an example of how to convert a view as part of the upgrade path.

**Comment by klonos** [comment]
Thought that I could give this a go. While tackling #549, I noticed that we already have a "Recent comments" view but it is disabled. So, what's left to be done here?

**Comment by klonos** [comment]
Thought that what needs to be done here is remove the code that provides the non-view "recent comments" block and then write an update hook that:

1. Enables the "recent comments" view.
2. Recursively goes through each layout and checks if the non-views provided "recent comments" block is placed in any of them.
3. For any layout that does contain the non-views provided "recent comments" block, we need to replace it with the one provided by views. In the process, copy/retain any settings that apply for the views block.

**Comment by klonos** [comment]
...

4. Minor problem (if I got this correctly): We can have multiple "Recent comments" blocks provided by comment.module in various layouts/regions. Each one of these recent comments blocks might have different settings. All the block-related settings (block title, style settings and visibility conditions) can be easily retained when replacing these blocks with their views-provided counterparts, but the "Number of recent comments" setting can be set individually for each block when the blocks are provided by comment.module (it is a block-level setting saved individually per-block instance), while with the views-provided "Recent comments" blocks that setting is "pulled" from the view (it is actually the view pager setting). If we want different setting per block instance, then the only way I see this being possible is to create a separate block display for each block that needs to have 

5. Related to the previous point above... the default number of recent comments for blocks provided by comment.module is 10 while the view has a pager of 5 items set in the default display. We can change the default display (and the block display will automatically inherit the new number), but if it so happens that this display is actually used elsewhere in the website, the we might mess it up. Again, the best approach in my opinion is to handle this with separate block displays for the "Recent comments" view and by setting the pager individually for each display, independently from the master/default display.

**Comment by jenlampton** [comment]
>  If we want different setting per block instance, then the only way I see this being possible is to create a separate block display for each block that needs to have.

There is a setting that was available on views (content pane displays) in Drupal 7 but was not ported to Backdrop.  The view should have a display-setting that allows per-instance configurations (on things like number of items, field displays, contextual filter sources, etc).

Do we have an issue for "Make views block displays more like views content pane displays" or similar?  This issue might be blocked by that one...  Also related, the "exposed filters don't work w/o ajax" bug.

**Comment by klonos** [comment]
Hmm, never heard of *content pane displays* before. Looking into that...

**Comment by klonos** [comment]
OK, an initial (mostly untested) PR up for review. This is what I have so far...

- New update hook in comment.install (did I choose wisely, or should the update hook be placed elsewhere?)
- Checks if the "Recent comments" view is enabled and if not it enables it.
- Goes through all layouts and gets all "Recent comments" blocks provided by the comment.module.
- If there were no "Recent comments" blocks found, nothing else happens. People can use the block provided by the "Recent comments" view (still need to remove the "Recent comments" block provided by comment.module).
- If there were "Recent comments" blocks found, it creates a new block display in the "Recent comments" view for each one of them and sets the pager to the same number as the "Number of recent comments" setting.
- It then alters all the old "Recent comments" blocks from being provided by the comment.module to being provided by the "Recent comments" view instead. Specifically, each block is mapped to a separate block display of the view. This makes it possible to account for cases where there were various blocks placed in the site with different "Number of recent comments" settings. Once we have #2433 in core and it is possible to have blocks provided by a single view block display but still being able to have different settings, then we can update all these blocks to point to the same block display of that "Recent comments" view.

I'm sure that 80% of the sites out there will not have more than a couple of "Recent comments" blocks, so this should not mess things up too much (like ending up with dozens of block displays in the "Recent comments" view).

**Comment by jenlampton** [comment]
I wonder about replacing the current "recent comments" block view, with the
new "admin comments" page view. I expect most people will need the page but
prefer to leave the block view disabled (an admin-only block is a weird
thing...Leftovers from D7.).

I would probably add a new view that replaces the old page, but leave the
recent comments block alone. That will make for a cleaner upgrade path too,
and make it easier to remove the recent comments block view in 2.x (which I
think was the plan?)

- -
Please excuse brevity and typos. Message typed with my thumbs from an Open
Source OS.

**Comment by klonos** [comment]
> I wonder about replacing the current "recent comments" block view, with the new "admin comments" page view.

Well, we'll be calling it "Manage comments" view 😉 (see my incomplete [PR](https://github.com/backdrop/backdrop/pull/1691) for #549), but anyways, I don't think that mixing admin views with content-presenting views is such a good idea. For one it's simply a matter of naming them intuitively in order for people to be able to easily tell which view does what and which .json config file is the one that goes with which view. If you have a view called "Manage comments", then people can easily associate that with the respective admin UI page and it will be consistent with the "Manage content", "Manage files" and "Manage users" views and the respective admin pages and .json config files. Replacing the "Recent comments" view with a block display of the "Manage comments" view will make things less intuitive and self-explanatory. We assume that people will be successfully making the association that a "Manage comments" view (with a config file named `views.view.manage_comments.json`) also holds the "Recent comments" block, but that might not be the case and it might result in more WTFish DX.

> I would probably add a new view that replaces the old page, but leave the recent comments block alone.

Not sure I follow you here (which view are you referring to as new and to which as old)?

> That will make for a cleaner upgrade path too, and make it easier to remove the recent comments block view in 2.x (which I think was the plan?)

Well, with #2433 implemented, we'd have the "Recent comments" blocks provided by the "Recent comments" view. We'd simply have to adapt the block display of that view to include the new features of the views content panes (in other words expose the pager settings that were previously set within the view as block settings instead). Removing that view defeats the purpose, so that's the opposite of the plan 😄

**Comment by jenlampton** [comment]
Maybe I am getting two issues confused. I thought this was the 'manage
comments' page view issue. Looks like that's not the case! Sorry, please
ignore.

- -
Please excuse brevity and typos. Message typed with my thumbs from an Open
Source OS.

**Comment by quicksketch** [comment]
I reviewed the PR at https://github.com/backdrop/backdrop/pull/1693#pullrequestreview-14809967. Overall a great job understanding the problem and providing a workable solution! However, I do think we should postpone this on the new issue you created about making things like this overriddable on a per-block basis in #2433.

Even so, the PR forms a solid base for iterating in the future, and you've covered all the situations that need to be addressed.

**Comment by klonos** [comment]
Fair enough @quicksketch. I prefer to get things right in the first place when possible and in this case it does mean getting #2433 in core first otherwise there's chances that it will be messy.

**Comment by klonos** [comment]
...I hope it's ok if I leave the PR as is and do not work on the changes requested in the review. It would be a waste of time before #2433 happens and a lot could change in the meantime.
